### PR TITLE
macOS font selection improvement (see #977)

### PR DESCRIPTION
**Pull request by chearon**
This makes the font example work on macOS.

People can still run into the Pango bug on Mac if they do something like `ctx.font = "10px Arial, Helvetica".`

Behdad said that the patch looked good but hasn't merged it yet: https://bugzilla.gnome.org/show_bug.cgi?id=762873

### CODE REVIEW

**Comment by LinusU** [comment]
:beers:

File: src/Canvas.cc [hunk 1]
```diff
@@ -750,8 +751,14 @@ Canvas::ResolveFontDescription(const PangoFontDescription *desc) {
 
     for (; it != _font_face_list.end(); ++it) {
       if (g_ascii_strcasecmp(families[i], pango_font_description_get_family(it->user_desc)) == 0) {
-        if (renamed_families->len) g_string_append(renamed_families, ",");
-        g_string_append(renamed_families, pango_font_description_get_family(it->sys_desc));
+        char *name = g_strdup(pango_font_description_get_family(it->sys_desc));
+
+        // Avoid sending duplicate SFNT font names due to a bug in Pango for macOS:
+        // https://bugzilla.gnome.org/show_bug.cgi?id=762873
+        if (g_hash_table_add(seen_families, name)) {
```
Comment (zbjornson) [inline src/Canvas.cc]: @chearon `g_hash_table_add` was added in 2.32, and the latest Windows GTK bundle (the one suggested in the install instructions) is 2.22.

Apparently it's equivalent to `g_hash_table_replace(seen_families, name, name)`, but in 2.22 `g_hash_table_replace` returns `void`. Also 2.22 has no `_contains` function, so maybe this:

```cpp
        bool existed = g_hash_table_lookup(seen_families, name) == NULL;
        g_hash_table_replace(seen_families, name, name);
        if (existed) {
```

That fixes the Windows build, but I don't have a Mac to test if it fixes the font resolution bug.

(2.22 source: https://github.com/GNOME/glib/blob/glib-2-22/glib/ghash.c)

Comment (chearon) [reply src/Canvas.cc]: Ugh, 2.22 was released in 2009. 2.32 in 2012

I've also thought of maybe switching to c++ libraries, but the GLib stuff plays nicely with Pango.

I'll switch it to `g_hash_table_replace` in a sec

**Comment by LinusU** [comment]
Published in `2.0.0-alpha.5`

**Pull request merged**